agent/wsclient/client.go

```diff
@@ -230,15 +223,51 @@ func (cs *ClientServerImpl) IsReady() bool {
 
 // SetConnection passes a websocket connection object into the client. This is used only in
 // testing and should be avoided in non-test code.
-func (cs *ClientServerImpl) SetConnection(conn WebsocketConn) {
+func (cs *ClientServerImpl) SetConnection(conn wsconn.WebsocketConn) {
 	cs.conn = conn
 }
 
 // SetReadDeadline sets the read deadline for the websocket connection
 // A read timeout results in an io error if there are any outstanding reads
 // that exceed the deadline
 func (cs *ClientServerImpl) SetReadDeadline(t time.Time) error {
-	return cs.conn.SetReadDeadline(t)
+	err := cs.conn.SetReadDeadline(t)
+	if err == nil {
+		return nil
+	}
+	seelog.Warnf("Unable to set read deadline for websocket connection: %v for %s", err, cs.URL)
+	// If we get connection closed error from SetReadDeadline, break out of the for loop and
+	// return an error
+	if opErr, ok := err.(*net.OpError); ok && strings.Contains(opErr.Err.Error(), errClosed) {
+		seelog.Errorf("Stopping redundant reads on closed network connection: %s", cs.URL)
+		return opErr
+	}
+	// An unhandled error has occurred while trying to extend read deadline.
+	// Try asynchronously closing the connection. We don't want to be blocked on stale connections
+	// taking too long to close. The flip side is that we might start accumulating stale connections.
```

Comment:
mainly for my understanding - what exactly happens if we "start accumulating stale connections"?

Reply:
at some point of time, the OS is going to stop the ECS agent from creating new connections if there's a chronic issue because of file descriptor limits.

Reply:
How do we know that? Do we have a test that ensures that the expected behavior happens when we hit the limit?